### ISSUE TITLE
KeyPair struct?

### ISSUE THREAD

**Issue by D4nte** [rust-bitcoin/rust-secp256k1#149]
We have created a `KeyPair` struct that is practical to manipulate secp256k1 key pairs. See https://github.com/comit-network/comit-rs/blob/master/vendor/secp256k1_support/src/keypair.rs

Would you be interested in us contributing it back to this repository? As this is a Rust binding project I am not sure it is the correct way forward.

**Comment by apoelstra** [comment]
Thanks! My feeling is probably not - in my experience it's pretty rare that any code works with both private keys and public keys at the same time.

**Comment by D4nte** [comment]
Ok, I'll see if it's an unnecessary optimisation in our codebase.

**Comment by thomaseizinger** [comment]
> Thanks! My feeling is probably not - in my experience it's pretty rare that any code works with both private keys and public keys at the same time.

It might be rare but if it is the case, the current situation is quite cumbersome.
The reason is that, in order to compute the `PublicKey` from the `SecretKey`, you need to have a reference to the `Secp256k` context. That one should obviously not be created on-the-fly and hence you either need a global instance (through lazy_static for example) or pass it in as a parameter. Neither of those two options are very pleasant.

I don't think a KeyPair struct would be very invasive in `rust-secp256k1`. However, it would make the situations in which some code needs to have both available much more pleasant to work with.

One could also pass around a tuple but that doesn't give you any guarantees that the keys are not arbitrarily combined, which would be a nice side-effect of a `KeyPair` struct (given proper field visibility).

**Comment by elichai** [comment]
@thomaseizinger what's the difference between having this struct in the lib and the user writing it?

In my experience people write this sort of KeyPair struct to abstract out the exact library from their logic, such that in theory they could "easily" swap it out.

Which makes some sense in combination with lazy_static/sync::Once

**Comment by thomaseizinger** [comment]
> @thomaseizinger what's the difference between having this struct in the lib and the user writing it?
> 
> In my experience people write this sort of KeyPair struct to abstract out the exact library from their logic, such that in theory they could "easily" swap it out.
> 
> Which makes some sense in combination with lazy_static/sync::Once

No big difference :). In our concrete case, we started our project as a single repository with several crates using path dependencies. We now try to sort that out more and more and the `KeyPair` struct is something that is referenced in several places which we would (ideally) like to be independent crates. Hence, if it is in `secp256k1`, we don't have to publish our own crate that contains such struct that we can re-use across our crates.

**Comment by apoelstra** [comment]
I don't understand the ergonomic benefit of this. Your `From` impl means that arbitrary combinations of secret/public keys are still possible; your `new_random` function requires a `rand` dependency; and both `new_random` and `from_secret_key` still require a context object.

**Comment by thomaseizinger** [comment]
> Your From impl means that arbitrary combinations of secret/public keys are still possible

I purposely left out the implementation that would allow this (`impl From<(SecretKey, PublicKey)> for KeyPair`) and only added the one that allows you to deconstruct it into a tuple: `impl From<KeyPair> for (SecretKey, PublicKey)`

> your new_random function requires a rand dependency; and both new_random and from_secret_key still require a context object.

Yes, obviously the `KeyPair` has to be constructed somehow. The `new_random` function guarantees that:

1. It is a random
2. It uses the secp-context to do that

> I don't understand the ergonomic benefit of this.

In a nutshell, the ergonomic improvement is to generate a `KeyPair` in one part of the application and passing it around (possibly to code in another crate) that doesn't have/want access to an secp-context while maintaining a compile-time guarantee that `SecretKey` and `PublicKey` belong together. This is not given by passing around a tuple.

Now, any project is free to create such a type as correctly pointed out by @elichai. All I am saying is that it can be added in a non-invasive way and it allows users to build onto that compile-time guarantee.

**Comment by thomaseizinger** [comment]
After discussing this in the team, we concluded that adding the `Keypair` here would actually not be enough for our usecase.
Concretely, we discovered that even given a `Keypair`, we would still need access to the context for operations like signing so there would be no added benefit.

Hence, we will very likely provide our own crate on top of `secp256k1` that provides this functionality.

Thanks for considering it though :)
